Add option for list filter quick remove

merwok opened the pull request:
fixes #181

merwok commented:
![dai-active-filters](https://user-images.githubusercontent.com/635179/194569250-7094f872-58b1-4099-9a8b-c0b93d046fc1.png)

merwok commented:
It looks a bit silly for a page with only one or two filters, but useful for a list with many filters.

Maybe we need a way to control theme settings per model instead of globally per project, but that’s a big change!

merwok commented on admin_interface/static/admin_interface/css/admin-interface.css:
```diff
@@ -376,6 +376,15 @@
+    font-size: 13px;
+    margin: .3em 0;
+    padding: 0 15px;
+    text-transform: capitalize;
```
```suggestion

```

merwok commented on admin_interface/templatetags/admin_interface_tags.py:
```diff
@@ -88,3 +94,15 @@ def hash_string(text):
+def admin_interface_filter_display(changelist, list_filter):
+    title = list_filter.title
+    value = next(c for c in list_filter.choices(changelist) if c['selected'])
+    return '{}: {}'.format(title, value['display'])
```
```suggestion
    return '{}: {}'.format(title.capitalize(), value['display'])
```

merwok commented on admin_interface/models.py:
```diff
@@ -340,6 +340,9 @@ def get_active_theme():
         default=True, verbose_name=_("sticky position")
     )
+    list_filter_quick_remove = models.BooleanField(
+        default=True, verbose_name=_("quick remove")
```
Should this help text be `quick remove links` ?  `quick clear links at top of sidebar` ?

Or short text is ok if README explains it with a full sentence and a screenshot?

fabiocaccamo replied:
I think that the right name for this option is `list_filter_chips` (chips is the technical name for this kind of control, generally used with tags).

The `help_text` could be `"quick remove active filters at top of sidebar"`.

merwok replied:
One issue is that `chips` is about the presentation, not the functionality.  What do you think of the compromise `list_filter_removal_chips` ?

fabiocaccamo replied:
I like it!

merwok replied:
I had forgotten that I am not proposiing chips any more!

From my first comment here:
> In the end I applied simple styling instead of the «chip» idea.

This was both for a practical reason (easier to start from the existing styling on `li` and adjust than add pill styling to `a`) and a UI reason: when you see chips for example in a many-to-many autocomplete field form, you expect to be able to delete more than one in a row, because they are little informative + interactive elements, not links that refresh the whole page and make you wait.

fabiocaccamo replied:
So... `list_filter_removal_links`?

merwok commented on admin_interface/templatetags/admin_interface_tags.py:
```diff
@@ -65,15 +65,21 @@ def get_admin_interface_languages(context):
-@simple_tag(takes_context=True)
-def get_admin_interface_theme(context):
+@simple_tag()
+def get_admin_interface_theme():
```
The function is not using context, so I removed the parameter to make it easier to call from the next function.

Is there something I don’t understand about it?  There was a unit test related to context but the assertions were the same as for the test without so I don’t get why this tag was taking context.

fabiocaccamo replied:
Frankly I don't remember this, probably it is not needed at all and it is caused by a copy/paste.

fabiocaccamo reviewed (changes requested):
- I would highlight the active filters option value, making it bold or darker, eg. 
(Status: **Published** x)

- If an active filter value is too long, the text overflow should be hidden and the text truncated with ellipsis.

Optional up to you:

- I'm not really convinced by the minimal style, but maybe the pill control would be too much in case of multiple tags. Just an idea: giving the possibility too choose also the style for the active filters like for the `language_chooser`?

- As the style for the `list_filter` customizations is always more, maybe it's a good idea to move it to its own CSS file. It will be easier to maintain. There is already a CSS file for the [`list_filter_dropdown`](https://github.com/fabiocaccamo/django-admin-interface/blob/master/admin_interface/static/admin_interface/css/list-filter-dropdown.css): maybe we can rename it to `list-filter.css` and place there all the style for the list filter, what you think?

merwok commented:
Changed model: new name and help text + default false.
It makes sense to have default true for a non-intrusive visual improvement like list_filter_highlight, but here it’s a UI change and it’s only useful if you have long list lists of filters in your admins.

Changed style: selected option in removal link is bold

Changed implementation: the template tag only prepares some values, leaving the actual rendering to a template, which projects can easily override!  So I can remove the bold if I don’t like it, and you can add truncation of selected option 🙂 

> Just an idea: giving the possibility too choose also the style for the active filters like for the language_chooser?

This can be tested with an overriden template in a project, and possibly brought into django-admin-interface with another option to select the template.  I don’t think I’ll have the time to implement it myself.

> As the style for the list_filter customizations is always more, maybe it's a good idea to move it to its own CSS file. It will be easier to maintain.

Because git still has terrible handling of tracking history of renamed files, I would not do this.
But you could do it before the next release, maybe making sure that each section of the file has a comment to explain to which feature it’s related.

fabiocaccamo reviewed (changes requested):
You forgot to add the updated migration.

fabiocaccamo commented on admin_interface/migrations/0027_theme_list_filter_removal_links.py:
```diff
@@ -0,0 +1,18 @@
+        migrations.AddField(
+            model_name='theme',
+            name='list_filter_removal_links',
+            field=models.BooleanField(default=False, verbose_name='quick remove links for active filters at top of sidebar'),
```
I like to make new features enabled by default `default=True`, users notice them and report possible issues. What do you think about it?

merwok replied:
> It makes sense to have default true for a non-intrusive visual improvement like list_filter_highlight, but here it’s a UI change and it’s only useful if you have long list lists of filters in your admins.

merwok commented on admin_interface/templates/admin/change_list.html:
```diff
@@ -0,0 +1,25 @@
+      {% block filters %}
+        {% if cl.has_filters %}
+          <div id="changelist-filter">
+            <h2>{% translate 'Filter' %}</h2>
```
I forgot to add translators comments to note these strings are already translated in the catalogs for django.contrib.admin!

fabiocaccamo replied:
Do you want to add them?
